### PR TITLE
Adding HA support

### PR DESCRIPTION
**Pull request by blublinsky**
Tested it on session cluster, usage for the application cluster is based on reply from Flink community

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @blublinsky,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by blublinsky** [comment]
I did sign the CLA.

**Comment by yuchaoran2011** [comment]
@blublinsky Could you provide some overview on the changes here? Are the newly added files adapted from the Flink code base? Is ZooKeeper a dependency for HA (I saw it mentioned several times)?

**Comment by blublinsky** [comment]
This is a completely new implementation, inspired by zookeeper implementation. The main things are described here https://github.com/lightbend/fdp-flink-build/blob/ha/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java.
It has no dependencies on zookeeper.
There are overwrites for 2 existing classes - HighAvailabilityServicesUtils, which adds support for filesystem and HighAvailabilityMode - adding filesystem as an option

**Comment by blublinsky** [comment]
And it is fully functional now, including application cluster

**Comment by deanwampler** [comment]
This is no longer a small wrapper around the existing Flink images with added Prometheus integration, right? As @yuchaoran2011 asked, it seems that extensive changes are being made and it's not clear what's changed, what the purpose of the changes is, and how well tested they are for production usage scenarios. A good discussion of what's new in the README would help.

**Comment by blublinsky** [comment]
It is a completely new implementation. So with the exception of 2 files mentioned above, its a new implementation.
Will add readme

**Comment by yuchaoran2011** [comment]
@blublinsky With your latest commits, the build was able to run. But it failed at the last step when building docker images. The error message is:
```
invalid argument "****/flink:.11_alpine" for "-t, --tag" flag: invalid reference format
See 'docker build --help'.
```
I wonder if it's because you don't have curly brackets around the env var FLINK_VERSION.

File: ci-build.sh [hunk 1]
```diff
@@ -1,63 +1,73 @@
...
+echo "metrics.reporters: prom" >> "${DISTRO}/scala12/conf/flink-conf.yaml"
+echo "metrics.reporter.prom.class: org.apache.flink.metrics.prometheus.PrometheusReporter" >> "${DISTRO}/scala12/conf/flink-conf.yaml"
+echo "metrics.reporter.prom.port: 9249" >>  "${DISTRO}/scala12/conf/flink-conf.yaml"
+
+docker build --build-arg flink_dist="${DISTRO}/scala11" -f Dockerfile_alpine -t lightbend/flink:$FLINK_VERSION_scala_2.11_alpine .
+docker build --build-arg flink_dist="${DISTRO}/scala12" -f Dockerfile_alpine -t lightbend/flink:$FLINK_VERSION_scala_2.12_alpine .
+docker build --build-arg flink_dist="${DISTRO}/scala11" -f Dockerfile_debian -t lightbend/flink:$FLINK_VERSION_scala_2.11_debian .
+docker build --build-arg flink_dist="${DISTRO}/scala12" -f Dockerfile_debian -t lightbend/flink:$FLINK_VERSION_scala_2.12_debian .
+docker build --build-arg flink_dist="${DISTRO}/scala11" -f Dockerfile_ubuntu -t lightbend/flink:$FLINK_VERSION_scala_2.11_ubuntu .
+docker build --build-arg flink_dist="${DISTRO}/scala12" -f Dockerfile_ubuntu -t lightbend/flink:$FLINK_VERSION_scala_2.12_ubuntu .
```
Comment (yuchaoran2011) [inline ci-build.sh]: It's these lines I'm talking about. Let's change all usages of $FLINK_VERSION to ${FLINK_VERSION}, as you did in other places.

Comment (blublinsky) [reply ci-build.sh]: Yep, you are right, just pushed update

File: Dockerfile_alpine [hunk 2]
```diff
@@ -1,21 +1,59 @@
...
+	; \
+	\
+	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
+	wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch"; \
+	wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc"; \
+	\
+# verify the signature
+	export GNUPGHOME="$(mktemp -d)"; \
+# for flaky keyservers, consider https://github.com/tianon/pgp-happy-eyeballs, ala https://github.com/docker-library/php/pull/666
+	gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4; \
```
Comment (deanwampler) [inline Dockerfile_alpine:22]: Should this key string be here?

File: ci-build.sh [hunk 3]
```diff
@@ -1,63 +1,73 @@
...
+
+curl ${DOWNLOAD_FLINK_URL} --output ${CURL_OUTPUT}
+
+# Untar source file and remove it
+
+tar xvzf ${CURL_OUTPUT} -C ${TMPDIR}
+rm ${CURL_OUTPUT}
+
+# Update code
+yes | cp -rf src/main/java/* ${TMPDIR}/flink-${FLINK_VERSION}/flink-runtime/src/main/java
```
Comment (deanwampler) [inline ci-build.sh:31]: !! This isn't a viable approach for something people should run in production, as it "hacks" the source code with a patch in an unusual way and then builds.

I think you need to refactor the HA support as follows:
1. Move the Flink code changes to a fork of the Flink project (which should be proposed as a PR to Flink), then build artifacts from that fork.
2. Have this build repo only do the minimum amount of work to build images from that fork with Prometheus and enabling this HA approach.

Comment (deanwampler) [reply ci-build.sh:31]: The Flink build needs a custom version string, too, like `1.8.0-lightbend-2.2.0` or something.

File: docker-entrypoint.sh [hunk 4]
```diff
@@ -32,21 +32,35 @@ elif [[ "${CMD}" = "${JOB_MANAGER}" || "${CMD}" = "${JOB_CLUSTER}" || "${CMD}" =
     shift
     sed -i -e "s/jobmanager.rpc.address: localhost/jobmanager.rpc.address: ${JOB_MANAGER_RPC_ADDRESS}/g" "$FLINK_HOME/conf/flink-conf.yaml"
     echo "metrics.internal.query-service.port: ${CONTAINER_METRIC_PORT}" >> "$FLINK_HOME/conf/flink-conf.yaml"
```
Comment (deanwampler) [inline docker-entrypoint.sh:34]: Bash tip: use this instead:

```bash
cat <<EOF >> "$FLINK_HOME/conf/flink-conf.yaml"
metrics.internal.query-service.port: ${CONTAINER_METRIC_PORT}
rest.port: 8081
blob.server.port: 6124
query.server.port: 6125
EOF
```

and similarly for the subsequent sections. I wouldn't change it now, though...

**Review by deanwampler — CHANGES_REQUESTED** [review]
I can't approve this as is for our use in Pipelines. The way Flink is patched here is not an acceptable approach for us to support in production scenarios. These changes need to be moved to a proper Flink fork and adequate tests provided. Those tests would be required for an upstream PR to be accepted, too.

Then this repo should remain small and focused on using builds of that fork (from Jenkins only!), adding Prometheus integration, and enabling this HA enhancement.

**Comment by deanwampler** [comment]
I'm closing this PR. As I said before, it's not acceptable to patch Flink itself this way. Instead, create an appropriate patch for Flink itself and remove those changes here, creating a new PR in this repo that uses those changes, as appropriate.

**Pull request closed without merging**